### DISCUSSION TITLE
Mini.files to toggle preview on demand

### DISCUSSION THREAD

**Discussion by carlesba** [nvim-mini/mini.nvim#936] [Ideas]
Would you consider adding a command to toggle preview panel?

**Comment by echasnovski** [comment]
This is already possible with `MiniFiles.config.windows.preview = not MiniFiles.config.windows.preview` command. It will take effect on the next refresh (usually when moving up/down/in/out).

**Comment by mau-mauricelim** [comment]
@echasnovski I did notice that it takes effect only on the next refresh as well.
@carlesba this is what I have currently that might help you! Not sure if that is the "right" way as I am still trying to learn lua but it works for my use-case.

I tried ``` vim.api.nvim_feedkeys('mzjkkj`z', 'n', false)``` but it did not work so well.

```
-- Create mapping to toggle preview
local show_preview = MiniFiles.config.windows.preview
local vert_move = function(is_top) return is_top and 'j' or 'k' end
local toggle_preview = function()
  show_preview = not show_preview
  MiniFiles.refresh({ windows = { preview = show_preview } })
  -- HACK: to force window to refresh, does not work with single dir/file items
  local is_top = vim.api.nvim_win_get_cursor(0)[1] == 1
  local first_move = vert_move(is_top)
  local last_move = vert_move(not is_top)
  vim.api.nvim_feedkeys(first_move, 'n', false)
  vim.defer_fn(function() vim.api.nvim_feedkeys(last_move, 'n', false) end, 0)
end
vim.api.nvim_create_autocmd('FileType', {
  pattern = { 'minifiles' },
  callback = function()
    vim.keymap.set('n', '<C-p>', toggle_preview, { buffer = true })
  end,
})

```

**Reply by echasnovski** [reply]
I'd suggest to not have all that, as it might introduce some unexpected behavior. My suggestion would be to just toggle with the `MiniFiles.config.windows.preview = not MiniFiles.config.windows.preview` and then manually move cursor. I don't think it should be a frequent occasion to toggle preview, so this should be fine.

**Comment by rbhanot4739** [comment]
I stumbled upon this thread and have something very similar in my config

```lua
vim.api.nvim_create_autocmd("User", {
  pattern = "MiniFilesBufferCreate",
  callback = function(args)
    vim.keymap.set("n", "<C-p>", function()
      MiniFiles = require("mini.files")
      MiniFiles.config.windows.preview = not MiniFiles.config.windows.preview
      MiniFiles.refresh({ windows = { preview = MiniFiles.config.windows.preview } })
    end, { desc = "toggle preview" })
  end,
})
```

This works but the only issue I see is that when preview is turned off the `MiniFiles.refresh` does not completely close the preview window until I press a key.


https://github.com/user-attachments/assets/c46747cf-688a-4467-867e-84fd2088cf76

**Reply by echasnovski** [reply]
Yes, it is mentioned in [this comment](https://github.com/echasnovski/mini.nvim/discussions/936#discussioncomment-9640297). Preview is expected to be either always on or always off during a single "explorer session". Being able to toggle it with a small caveat of having to press another key is considered a bonus, but not an intended use case.

**Reply by drowning-cat** [reply]
See https://github.com/nvim-mini/mini.nvim/discussions/2238

```lua
local toggle_preview = function()
  local is_enabled = MiniFiles.config.windows.preview
  local enable = not is_enabled
  MiniFiles.config.windows.preview = enable
  MiniFiles.trim_right()
  MiniFiles.refresh({ windows = { preview = enable } })
  MiniFiles.refresh()
  if is_enabled then
    local branch = MiniFiles.get_explorer_state().branch
    table.remove(branch)
    MiniFiles.set_branch(branch)
  end
end

vim.api.nvim_create_autocmd("User", {
  pattern = "MiniFilesBufferCreate",
  group = vim.api.nvim_create_augroup("minifiles.keymaps", { clear = true }),
  callback = function(args)
    local buf = args.data.buf_id
    vim.keymap.set("n", "<M-p>", toggle_preview, { buffer = buf, desc = "Toggle preview" })
  end,
})
```
